### PR TITLE
Component highlighter

### PR DESCRIPTION
**Pull request by KutnerUri**
## Proposed Changes

- Refactor bit react transformer to work with pgk.json in node_modules, add types, etc
- Move highlighter to be a provider in React Env
- Rewrite highlighter with React 16 hooks
- Compositions menu bar +slot
-

### CODE REVIEW

File: workspace.jsonc [hunk 1]
```diff
@@ -72,6 +75,8 @@
         "@teambit/base-ui.theme.fonts.roboto": "0.6.7",
         "@teambit/base-ui.theme.sizes": "0.6.7",
         "@teambit/base-ui.theme.theme-provider": "0.6.7",
+        "@teambit/base-ui.utils.popper-js.ignore-popper-size": "0.6.8",
+        "@teambit/base-ui.utils.popper-js.resize-to-match-reference": "0.6.8",
```
[the remark this answers is not in the document]
Comment (odedre) [reply workspace.jsonc:79]: I also saw that they use `export * from ...` which we agreed not to use anymore

File: scopes/compositions/compositions/ui/composition-preview.tsx [hunk 2]
```diff
@@ -15,19 +20,36 @@ export type ComponentCompositionProps = {
...
   hotReload?: boolean;
 };
 
-export function ComponentComposition({ component, composition, hotReload }: ComponentCompositionProps) {
+export function ComponentComposition({
+  component,
+  composition,
+  hotReload,
+  className,
+  queryParams = [],
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/compositions/compositions/ui/composition-preview.tsx:39]: ok but not right now

File: .bitmap [hunk 3]
```diff
@@ -1123,6 +1147,12 @@
         "mainFile": "index.ts",
         "rootDir": "scopes/ui-foundation/tree/tree-node"
     },
+    "ui/use-composition": {
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply .bitmap:1150]: ok

File: scopes/compositions/compositions/compositions.ui.runtime.tsx [hunk 4]
```diff
@@ -1,17 +1,28 @@
+import type { ReactNode } from 'react';
+import { Slot, SlotRegistry } from '@teambit/harmony';
 import { ComponentAspect, ComponentUI } from '@teambit/component';
 import { UIRuntime } from '@teambit/ui';
 
 import { CompositionsSection } from './composition.section';
 import { CompositionsAspect } from './compositions.aspect';
+import { MenuBarWidget } from './compositions';
+
+export type MenuBarWidgetsSlot = SlotRegistry<MenuBarWidget[]>;
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/compositions/compositions/compositions.ui.runtime.tsx:10]: renamed -> CompositionsMenuSlot

File: scopes/compositions/ui/compositions-menu-bar/compositions-menu-bar.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+interface CompositionsMenuBarProps extends React.HTMLAttributes<HTMLDivElement> {
+  menuBarWidgets?: MenuBarWidgetsSlot;
+}
+
+export function CompositionsMenuBar({ className, menuBarWidgets, ...rest }: CompositionsMenuBarProps) {
+  const widgetsStart = useMemo(
+    () =>
+      flatten(menuBarWidgets?.values())
+        .filter(({ location }) => location === 'start')
+        .map(({ content }, idx) => <React.Fragment key={idx}>{content}</React.Fragment>),
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/compositions/ui/compositions-menu-bar/compositions-menu-bar.tsx:16]: let's not make unnecessary `div`s until needed

File: scopes/compositions/ui/use-composition/use-composition.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    (value: T) =>
+      ctx.setQueryParams((state) => ({
+        ...state,
+        [paramName]: value,
+      })),
+    [paramName]
+  );
+
+  return [param as T, setQueryParam];
+}
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/compositions/ui/use-composition/use-composition.tsx:44]: ok

File: scopes/design/ui/input/option-button/option-button.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import classnames from 'classnames';
+import Button, { ButtonProps } from '@teambit/base-ui.input.button';
+import { Icon } from '@teambit/evangelist.elements.icon';
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/design/ui/input/option-button/option-button.tsx:4]: this is the only icon component until we migrate it to design

File: scopes/pkg/pkg/pkg.main.runtime.ts [hunk 8]
```diff
@@ -166,8 +166,8 @@ export class PkgMain {
    * returns the package path in the /node_modules/ folder
    */
   getModulePath(component: Component) {
-    const pgkName = this.getPackageName(component);
-    const path = join('node_modules', pgkName);
+    const pkgName = this.getPackageName(component);
+    const path = join('node_modules', pkgName);
     return path;
   }
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/pkg/pkg/pkg.main.runtime.ts:172]: not mine

File: scopes/react/bit-react-transformer/bit-from-pkg-json.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { existsSync, readFileSync } from 'fs';
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/bit-react-transformer/bit-from-pkg-json.ts:1]: ok

File: scopes/react/component-highlighter/component-highlighter/bit-react-component.tsx [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export interface BitComponent {
+  __bitComponentId: string;
+}
+
+export function isBitComponent(component: any): component is BitComponent {
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/component-highlighter/component-highlighter/bit-react-component.tsx:5]: ?

File: scopes/react/component-highlighter/component-highlighter/component-highlighter.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      setText(undefined);
+    }
+  }, [disabled]);
+
+  return (
+    <HoverSelector
+      {...rest}
+      onElementChange={handleElement}
+      disabled={disabled}
+      style={{ fontFamily: 'sans-serif' }}
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/component-highlighter/component-highlighter/component-highlighter.tsx:42]: we don't have our font in the compositions. so using the best sane defaults

File: scopes/react/component-highlighter/frame/frame.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  });
+
+  useEffect(() => {
+    const refocus = update;
+    if (!refocus || !motionTracking) return () => {};
+
+    let fid = 0;
+    const f = () => {
+      refocus().catch(() => {});
+      fid = window.requestAnimationFrame(f);
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/component-highlighter/frame/frame.tsx:58]: ok

File: scopes/react/component-highlighter/label/component-label/scope-bubble.tsx [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import classNames from 'classnames';
+
+import { ComponentID } from '@teambit/component-id';
+
+import styles from './duo-component-bubble.module.scss';
+
+export type ScopeBubbleProps = {
+  componentId: ComponentID;
+  fullScopeName?: boolean;
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/component-highlighter/label/component-label/scope-bubble.tsx:10]: ok. removed

File: scopes/react/component-highlighter/label/label.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
+import React, { useMemo, useState } from 'react';
+import { usePopper } from 'react-popper';
+import { ComponentID } from '@teambit/component-id';
+import type { CardProps } from '@teambit/base-ui.surfaces.card';
+import useAnimationFrame from 'use-animation-frame';
+import type { Placement, Modifier } from '@popperjs/core';
+import '@popperjs/core';
+
+import { DefaultLabel } from './default-lable';
+import { ComponentLabel } from './component-label';
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/component-highlighter/label/label.tsx:10]: yes, but in this PR

File: scopes/react/dom-to-react/dom-to-fiber.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+ * Return the nearest React Component Prototype containing the FiberNode
+ * @example
+ * const comp = fiberToPrototype(fiber);
+ * console.log(comp); // function LoginForm()
+ */
+export function fiberToPrototype(fiberNode: FiberNode | null) {
+  return toRootFiber(fiberNode)?.type || null;
+}
+
+// React 15 is not using Fiber, so support for it is more complicated.
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/dom-to-react/dom-to-fiber.ts:47]: then we try to use hooks and die. needs a separate highlighter for this

File: scopes/react/ui/highlighter-provider/highlighter-provider.tsx [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    updateHash();
+    window.addEventListener('hashchange', updateHash);
+
+    return () => {
+      window.removeEventListener('hashchange', updateHash);
+    };
+  }, []);
+
+  return hash;
+}
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/ui/highlighter-provider/highlighter-provider.tsx:36]: yes but not in this PR

File: workspace.jsonc [hunk 17]
```diff
@@ -37,11 +38,13 @@
         "@pnpm/store-connection-manager": "0.3.62",
         "@pnpm/tarball-fetcher": "8.2.8",
         "@pnpm/tarball-resolver": "4.0.7",
+        "@popperjs/core": "2.6.0",
         "@stencil/core": "1.17.4",
         "@stencil/webpack": "0.0.6",
         "@svgr/webpack": "5.5.0",
         "@teambit/any-fs": "0.0.5",
         "@teambit/base-ui.constants.storage": "0.6.3",
+        "@teambit/base-ui.css-components.pill": "0.6.7",
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply workspace.jsonc:47]: yes @odedre , border radius behaves differently with `px` and `%`. It's sets 11px horizontally and 50% vertically, resembling the shape of a pill

File: workspace.jsonc (hunk 1 above)
[the remark this answers is not in the document]
Comment (KutnerUri) [reply workspace.jsonc:79]: these are pure logic, with no styles

**Review by odedre — APPROVED** [review]

**Pull request merged**